docs/proposals/20250724-independent-worker-topology-versions.md

```diff
@@ -0,0 +1,515 @@
+to stagger upgrades or selectively upgrade only a subset of nodes. In the current model, this level of control is not possible 
+without abandoning ClusterClass altogether or creating multiple clusters—each of which increases operational complexity and cost.
+
+This proposal introduces an enhancement to support version skew between worker node groups in ClusterClass-managed clusters, enabling:
+
+    Safe, independent upgrades of individual worker node groups;
+
+    Reduced operational blast radius during upgrades;
+
+    Continued cluster operations (e.g., autoscaling, remediation) in unaffected worker node groups;
+
+    Greater flexibility for cluster vendors and multi-tenant platform teams who must balance upgrade safety with workload uptime guarantees.
+
+By decoupling worker node group upgrades from the global cluster version, we preserve the high-level abstraction and usability
+of ClusterClass while introducing operational flexibility needed for production environments at scale.
+
+The broader motivation aligns with the goals of Cluster API: declarative infrastructure, safe upgrades, and supporting real-world
+platform needs without requiring users to abandon upstream tools or adopt brittle workarounds.
+
+Optionally, future experience reports can be gathered from teams running large-scale Kubernetes fleets with mixed workloads
+(e.g., inference vs training, customer A/B rollout clusters) to further validate this direction.
+
+
+### Goals
+
+- Support upgrades of any single worker node group independent of the control plane and other worker node groups
+- Operations on worker node groups *not* upgrading should continue to operate normally while an independent worker node group is upgrading.
+- Support version skew, as allowed by Kubernetes' [Version Skew Policy](https://kubernetes.io/releases/version-skew-policy/), between worker node groups and control planes. 
+- Maintain supported skew checks and guardrails for Kubernetes versions
+- Maintain the ease of use of ClusterClass while providing this added flexibility
+- Gracefully handle [version-aware patches](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass#version-aware-patches) (i.e.  ClusterClass variables can reference `builtin.cluster.topology.version`, `builtin.controlPlane.version`, `builtin.machineDeployment.version`, and `builtin.machinePool.version`)
+- Update documentation in the [cluster-api book](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/operate-cluster#upgrade-a-cluster) to describe the alternative approach to upgrading cluster versions.
+
+### Non-Goals
+
+- Allow flip/flopping between cluster-level version control and independent worker version control to reduce potential edge cases
+- Allowing version skews that are unsupported by kubernetes
+- Changing behavior of cluster-level worker node group configuration rollouts
+- Add builtin version-aware patches for independent worker node groups
```

Comment:
what are builtin version-aware patches? I guess there's a link I should probably read first...